Objective-C/CBLDictionary.mm

```diff
@@ -264,28 +301,42 @@ - (BOOL) isEqual: (id)object {
     if (self.count != other.count)
         return NO;
     
-    for (MDict<id>::iterator i(_dict); i; ++i) {
-        NSString* key = i.nativeKey();
-        id value = i.nativeValue();
-        if (value) {
-            if (![value isEqual: [other valueForKey: key]])
-                return NO;
-        } else {
-            if ([other valueForKey: key] || ![other containsValueForKey: key])
-                return NO;
+    CBL_LOCK(self.sharedLock) {
+        for (MDict<id>::iterator i(_dict); i; ++i) {
+            NSString* key = i.nativeKey();
+            id value = i.nativeValue();
+            if (value) {
+                if (![value isEqual: [other valueForKey: key]])
+                    return NO;
+            } else {
+                if ([other valueForKey: key] || ![other containsValueForKey: key])
+                    return NO;
+            }
         }
     }
-    
     return YES;
 }
 
 
 - (NSUInteger) hash {
-    NSUInteger hash = 0;
-    for (MDict<id>::iterator i(_dict); i; ++i) {
-        hash += ([i.nativeKey() hash] ^ [i.nativeValue() hash]);
+    CBL_LOCK(self.sharedLock) {
+        NSUInteger hash = 0;
+        for (MDict<id>::iterator i(_dict); i; ++i) {
+            hash += ([i.nativeKey() hash] ^ [i.nativeValue() hash]);
+        }
+        return hash;
     }
-    return hash;
+}
+
+
+#pragma mark - Lock
+
+
+- (NSObject*) sharedLock {
+    CBLDatabase* db = nil;
```

Comment:
Same comment as in CBLArray: Precompute this in the init method and store it in an ivar; it's too expensive to compute on every property access. (Probably needs to be marked as `__weak` to avoid a ref-cycle.)